providers/mlx5/verbs.c

```diff
@@ -4461,30 +4635,43 @@ struct mlx5dv_devx_uar *mlx5dv_devx_alloc_uar(struct ibv_context *context,
 
 	int ret;
 	struct mlx5_bf *bf;
+	struct mlx5_context *ctx = to_mctx(context);
 
 	if (!is_mlx5_dev(context->device)) {
 		errno = EOPNOTSUPP;
 		return NULL;
 	}
 
-	if (flags) {
-		errno = ENOTSUP;
+	if (!check_comp_mask(flags, MLX5_IB_UAPI_UAR_ALLOC_TYPE_NC)) {
+		errno = EOPNOTSUPP;
 		return NULL;
 	}
 
-	bf = mlx5_attach_dedicated_bf(context);
-	if (!bf)
-		return NULL;
+	bf = mlx5_attach_dedicated_uar(context, flags);
+	if (!bf) {
+		if (flags || !(ctx->flags & MLX5_CTX_FLAGS_NO_KERN_DYN_UAR))
+			return NULL;
 
-	fill_attr_in_uint32(cmd, MLX5_IB_ATTR_DEVX_QUERY_UAR_USER_IDX,
-			    bf->bfreg_dyn_index);
-	fill_attr_out_ptr(cmd, MLX5_IB_ATTR_DEVX_QUERY_UAR_DEV_IDX,
-		      &bf->devx_uar.dv_devx_uar.page_id);
+		/* fallback to legacy mode */
+		bf = mlx5_attach_dedicated_bf(context);
```

Comment:
Why is this fallback here when the change to mlx5_alloc_td just switches mlx5_attach_dedicated_bf() to mlx5_attach_dedicated_uar() ?

Reply:
The TD is moved to use mlx5_attach_dedicated_uar() in the next patch upon adapting legacy mode to new one, by that step the fallback is really done internally.

Reply:
So is it deleted in a later patch? It looks weird

Reply:
Yes,  I preferred this incremental mode as of the massive cleanup/refactor in the next patch for the legacy mode to clarify things.